src/crimson/os/seastore/onode_manager/simple-fltree/onode_node.h

```diff
@@ -0,0 +1,962 @@
+    return std::equal(
+        lhs.begin(), lhs.end(),
+        rhs.begin(), rhs.end());
+  }
+}
+
+// on-disk onode
+// it only keeps the bits necessary to rebuild an in-memory onode
+struct [[gnu::packed]] onode_t {
+  onode_t& operator=(const onode_t& onode) {
+    len = onode.len;
+    std::memcpy(data, onode.data, len);
+    return *this;
+  }
+  size_t size() const {
+    return sizeof(*this) + len;
+  }
+  OnodeRef decode() const {
+    return new crimson::os::seastore::Onode(std::string_view{data, len});
+  }
+  uint8_t struct_v = 1;
+  uint8_t struct_compat = 1;
+  // TODO:
+  // - use uint16_t for length, as the size of an onode should be less
+  //   than a block (16K for now)
+  // - drop struct_len
+  uint32_t struct_len = 0;
+  uint32_t len;
+  char data[];
+};
+
+static inline std::ostream& operator<<(std::ostream& os, const onode_t& onode) {
+  return os << *onode.decode();
+}
+
+using crimson::os::seastore::laddr_t;
+
+struct [[gnu::packed]] child_addr_t {
+  laddr_t data;
+  child_addr_t(laddr_t data)
+    : data{data}
+  {}
+  child_addr_t& operator=(laddr_t addr) {
+    data = addr;
+    return *this;
+  }
+  laddr_t get() const {
+    return data;
+  }
+  operator laddr_t() const {
+    return data;
+  }
+  size_t size() const {
+    return sizeof(laddr_t);
+  }
+};
+
+// poor man's operator<=>
+enum class ordering_t {
+  less,
+  equivalent,
+  greater,
+};
+
+template<class L, class R>
+ordering_t compare_element(const L& x, const R& y)
+{
+  if constexpr (std::is_arithmetic_v<L>) {
+    static_assert(std::is_arithmetic_v<R>);
+    if (x < y) {
+      return ordering_t::less;
+    } else if (x > y) {
+      return ordering_t::greater;
+    } else {
+      return ordering_t::equivalent;
+    }
+  } else {
+    // string_view::compare(), string::compare(), ...
+    auto result = x.compare(y);
+    if (result < 0) {
+      return ordering_t::less;
+    } else if (result > 0) {
+      return ordering_t::greater;
+    } else {
+      return ordering_t::equivalent;
+    }
+  }
+}
+
+template<typename L, typename R>
+constexpr ordering_t tuple_cmp(const L&, const R&, std::index_sequence<>)
+{
+  return ordering_t::equivalent;
+}
+
+template<typename L, typename R,
+         size_t Head, size_t... Tail>
+constexpr ordering_t tuple_cmp(const L& x, const R& y,
+                               std::index_sequence<Head, Tail...>)
+{
+  auto ordering = compare_element(std::get<Head>(x), std::get<Head>(y));
+  if (ordering != ordering_t::equivalent) {
+    return ordering;
+  } else {
+    return tuple_cmp(x, y, std::index_sequence<Tail...>());
+  }
+}
+
+template<typename... Ls, typename... Rs>
+constexpr ordering_t cmp(const std::tuple<Ls...>& x,
+                         const std::tuple<Rs...>& y)
+{
+  static_assert(sizeof...(Ls) == sizeof...(Rs));
+  return tuple_cmp(x, y, std::index_sequence_for<Ls...>());
+}
+
+enum class likes_t {
+  yes,
+  no,
+  maybe,
+};
+
+struct [[gnu::packed]] variable_sized_key_t {
```

Comment:
This isn't generic.  I think you should rename this ghobject_binary_t or something.

Reply:
i'd prefer keeping this name, because

1. onode manager is all about mapping ghobject to onode. and `onode_node.{h,cc}` tries to offer the primitives for the indexing. in this context, i hope the readers of this source file bear in mind that "key" is always oid or part of it. just like we don't always put "ceph" as the prefix of our data structures in this project. so 
1. suffix is not quite right. as the key here is composed of three parts "fixed key", "variable sized key", yet another "fixed key". that last "fixed key" is snap and gen.
1. it matches the name in your design doc. i think it helps the reader to understand the code if he/she have read the design doc.

Reply:
Ok